packages/@aws-cdk/aws-codepipeline/lib/stage.ts

```diff
@@ -105,69 +90,100 @@ export class Stage extends cdk.Construct implements cpapi.IStage, cpapi.IInterna
+  public render(): any {
     return {
-      name: this.node.id,
+      name: this.stageName,
       actions: this._actions.map(action => this.renderAction(action)),
     };
   }
 
+  public addAction(action: cpapi.Action): cpapi.IStage {
+    // check for duplicate Actions and names
+    if (this._actions.find(a => a.actionName === action.actionName)) {
+      throw new Error(`Stage ${this.stageName} already contains an Action with name '${action.actionName}'`);
+    }
+
+    this._actions.push(action);
+    if (this.stageIsAttachedToPipeline()) {
+      this.attachActionToPipeline(action);
+    }
+
+    return this;
+  }
+
   public onStateChange(name: string, target?: events.IEventRuleTarget, options?: events.EventRuleProps) {
-    const rule = new events.EventRule(this, name, options);
+    const rule = new events.EventRule(this.getParent('onStateChange'), name, options);
     rule.addTarget(target);
     rule.addEventPattern({
       detailType: [ 'CodePipeline Stage Execution State Change' ],
       source: [ 'aws.codepipeline' ],
-      resources: [ this.pipeline.pipelineArn ],
+      resources: [ this.pipeline!.pipelineArn ],
       detail: {
-        stage: [ this.node.id ],
+        stage: [ this.stageName ],
       },
     });
     return rule;
   }
 
-  // can't make this method private like Pipeline#_attachStage,
-  // as it comes from the IStage interface
-  public _attachAction(action: cpapi.Action): void {
-    // _attachAction should be idempotent in case a customer ever calls it directly
-    if (!this._actions.includes(action)) {
-      this._actions.push(action);
+  protected validate(): string[] {
+    return this.validateHasActions();
+  }
 
-      (this.pipeline as any)._attachActionToRegion(this, action);
+  // ignore unused private method (it's actually used in Pipeline)
+  // @ts-ignore
+  private _attachStageToPipeline(pipeline: Pipeline): void {
+    if (this.stageIsAttachedToPipeline()) {
+      throw new Error(`Stage '${this.stageName}' has been added to a Pipeline twice`);
+    }
+    this.pipeline = pipeline;
+    this.parent = new cdk.Construct(pipeline, this.stageName);
+    for (const action of this._actions) {
+      this.attachActionToPipeline(action);
     }
   }
 
-  public _generateOutputArtifactName(action: cpapi.Action): string {
-    return (this.pipeline as any)._generateOutputArtifactName(this, action);
+  private stageIsAttachedToPipeline() {
+    return this.pipeline !== undefined;
   }
 
-  public _findInputArtifact(action: cpapi.Action): cpapi.Artifact {
-    return (this.pipeline as any)._findInputArtifact(this, action);
+  private attachActionToPipeline(action: cpapi.Action) {
+    const actionParent = new cdk.Construct(this.parent!, action.actionName);
+    (action as any)._attachActionToPipeline(this.pipeline, this, actionParent);
   }
 
-  protected validate(): string[] {
-    return this.validateHasActions();
+  private getParent(methodName: string): cdk.Construct {
```

Comment:
Rename parent to scope and change to property. The name of the method is available in the stack trace

Reply:
Done. I did leave the `methodName` though, like I discussed above (this is a customer-facing error).

Reply:
See comment above, I don't see a value here for customers. Stack traces are part of the error

Reply:
Done.